### PR TITLE
Fast binary conversion

### LINKED ISSUE #1468: Make `fromℕ` in `Data.Nat.Binary` more efficient
See #1405 of the discussion of the approaches. @oisdk's approach looks promising.

### PR DESCRIPTION
**Pull request by jmlowenthal**
Takes @oisdk's approach from the discussion in #1405 and cleans it up a bit, trying to integrate better with other parts of stdlib.

Feedback welcome, especially if there are other proofs in stdlib that could support in `fromℕ-helper≡fromℕ'`.

Closes #1468.

### CODE REVIEW

File: src/Data/Nat/Binary/Properties.agda [hunk 1]
```diff
@@ -98,15 +103,75 @@ toℕ-pred zero     =  refl
...
+
+  open ≡-Reasoning
+
+  tail-homo : ∀ n → tail (suc (fromℕ' n)) ≡ fromℕ' (n ℕ÷./ 2)
+  tail-homo ℕ.zero = refl
+  tail-homo (ℕ.suc ℕ.zero) = refl
+  tail-homo (ℕ.suc (ℕ.suc n)) = begin
+    tail (suc (fromℕ' (ℕ.suc (ℕ.suc n))))  ≡˘⟨ tail-suc (fromℕ' n) ⟩
+    suc (tail (suc (fromℕ' n)))            ≡⟨ cong suc (tail-homo n) ⟩
+    fromℕ' (ℕ.suc (n ℕ÷./ 2))              ≡⟨ cong fromℕ' (sym (divₕ-extractAcc 1 1 n 1)) ⟩
```
Comment (MatthewDaggitt) [inline src/Data/Nat/Binary/Properties.agda:158]: `Data.Nat.DivMod.Core` shouldn't be imported outside of `Data.Nat.DivMod`. Probably better to extract the result to it's own top-level proof in `Data.Nat.DivMod`!

File: src/Data/Nat/Binary/Properties.agda [hunk 2]
```diff
@@ -12,11 +12,16 @@ open import Algebra.Bundles
 open import Algebra.Morphism.Structures
 import Algebra.Morphism.MonoidMonomorphism as MonoidMonomorphism
 open import Algebra.Consequences.Propositional
+open import Data.Bool.Base using (if_then_else_; Bool; true; false)
+open import Data.Maybe.Base using (Maybe; just; nothing)
 open import Data.Nat.Binary.Base
 open import Data.Nat as ℕ using (ℕ; z≤n; s≤s)
+import Data.Nat.DivMod as ℕ÷
```
Comment (MatthewDaggitt) [inline src/Data/Nat/Binary/Properties.agda:19]: I'd probably just import this as `ℕ` rather than `ℕ÷`

File: src/Data/Nat/Binary/Properties.agda [hunk 3]
```diff
@@ -12,11 +12,16 @@ open import Algebra.Bundles
...
+open import Data.Nat.DivMod.Core using (divₕ-extractAcc)
 import Data.Nat.Properties as ℕₚ
 open import Data.Nat.Solver
-open import Data.Product using (_,_; proj₁; proj₂; ∃)
+open import Data.Product using (_×_; _,_; proj₁; proj₂; ∃)
+import Data.Product.Properties as Productₚ
```
Comment (MatthewDaggitt) [inline src/Data/Nat/Binary/Properties.agda:24]: Not used?

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]
Thank you for the awesome PR! I've added some comments, and moved `fromN-helper` method inside a local `where module` to limit it's outside exposure and make it clear that it's not supposed to be used on it's own.

Apart from the comments, the only other thing is to add CHANGELOG entries for any new proofs you've added.

File: src/Data/Nat/Binary/Properties.agda [hunk 4]
```diff
@@ -98,15 +103,75 @@ toℕ-pred zero     =  refl
...
+  where
+  split : ℕᵇ → Maybe Bool × ℕᵇ
+  split zero     = nothing , zero
+  split 2[1+ n ] = just false , n
+  split 1+[2 n ] = just true , n
+
+  head = proj₁ ∘ split
+  tail = proj₂ ∘ split
+
+  split-cong : ∀ m n → split m ≡ split n → m ≡ n
```
[the remark this answers is not in the document]
Comment (jmlowenthal) [reply src/Data/Nat/Binary/Properties.agda]: Updated to use `Injective`.

File: src/Data/Nat/Binary/Properties.agda [hunk 5]
```diff
@@ -98,15 +103,75 @@ toℕ-pred zero     =  refl
...
+  split-cong zero zero refl = refl
+  split-cong 2[1+ m ] 2[1+ .m ] refl = refl
```
[the remark this answers is not in the document]
Comment (jmlowenthal) [reply src/Data/Nat/Binary/Properties.agda]: Removed.

File: src/Data/Nat/Binary/Properties.agda (hunk 1 above)
Comment (jmlowenthal) [reply src/Data/Nat/Binary/Properties.agda]: Updated to use `+-distrib-/-∣ˡ` exported from `Data.Nat.DivMod`.

File: src/Data/Nat/Binary/Properties.agda (hunk 2 above)
Comment (jmlowenthal) [reply src/Data/Nat/Binary/Properties.agda]: Explicitly imported `_/_`, `_%_`, `m/n≤m` and `+-distrib-/-∣ˡ`.

File: src/Data/Nat/Binary/Properties.agda (hunk 3 above)
Comment (jmlowenthal) [reply src/Data/Nat/Binary/Properties.agda]: Removed.

**Comment by jmlowenthal** [comment]
Thank you for the prompt and useful feedback @MatthewDaggitt. I've been through and resolved all of your comments, all very sensible suggestions!

**Review by MatthewDaggitt — APPROVED** [review]

**Pull request merged**